### PR TITLE
backoff retry for cohere api, rate-limiting

### PR DESCRIPTION
**Pull request by mauriyouth**

### CODE REVIEW

File: core/src/providers/cohere.rs [hunk 1]
```diff
@@ -114,35 +114,29 @@ impl CohereLLM {
...
-                }
+        let response = match res.status() {
+            hyper::StatusCode::OK => {
+                let body = hyper::body::aggregate(res).await?;
+                let mut b: Vec<u8> = vec![];
+                body.reader().read_to_end(&mut b)?;
+                let c: &[u8] = &b;
+
+                let response: Response = serde_json::from_slice(c)?;
+                Ok(response)
```
[the remark this answers is not in the document]
Comment (spolu) [reply core/src/providers/cohere.rs:125]: To return non retryable errors

File: core/src/providers/cohere.rs [hunk 2]
```diff
@@ -114,35 +114,29 @@ impl CohereLLM {
...
             }
+            hyper::StatusCode::TOO_MANY_REQUESTS => Err(ModelError {
+                message: format!("CohereAPIError: {}", hyper::StatusCode::TOO_MANY_REQUESTS),
+                retryable: Some(ModelErrorRetryOptions {
+                    sleep: Duration::from_millis(2000),
+                    factor: 2,
+                    retries: 8,
+                }),
+            }),
+            status => Err(ModelError {
```
Comment (spolu) [inline core/src/providers/cohere.rs:135]: Sorry I presume the error body parsing should happen here

**Review by spolu — CHANGES_REQUESTED** [review]

File: core/src/providers/cohere.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mauriyouth) [reply core/src/providers/cohere.rs:125]: error parsing even api status code === 200  ?

[the remark this answers is not in the document]
Comment (spolu) [reply core/src/providers/cohere.rs:125]: No you’re right 👍

**Review by spolu — APPROVED** [review]
LGTM

**Pull request merged**